core/src/main/java/org/testcontainers/utility/DockerImageName.java

```diff
@@ -159,63 +176,86 @@ public String getRegistry() {
+     * behaves as the other does, and is compatible with Testcontainers' assumptions about the other image.
+     *
+     * @param otherImageName the image name of the other image
+     * @return an immutable copy of this {@link DockerImageName} with the compatibility declaration attached.
+     */
+    public DockerImageName asCompatibleSubstituteFor(DockerImageName otherImageName) {
+        return new DockerImageName(rawName, registry, repo, versioning, otherImageName);
     }
 
-    @Data
-    private static class Sha256Versioning implements Versioning {
-        public static final String HASH_REGEX = "[0-9a-fA-F]{32,}";
-        private final String hash;
-
-        Sha256Versioning(String hash) {
-            this.hash = hash;
+    /**
+     * Test whether this {@link DockerImageName} has declared compatibility with another image (set using
+     * {@link DockerImageName#asCompatibleSubstituteFor(String)} or
+     * {@link DockerImageName#asCompatibleSubstituteFor(DockerImageName)}.
+     * <p>
+     * If a version tag part is present in the <code>other</code> image name, the tags must exactly match, unless it
+     * is 'latest'. If a version part is not present in the <code>other</code> image name, the tag contents are ignored.
+     *
+     * @param other the other image that we are trying to test compatibility with
+     * @return whether this image has declared compatibility.
+     */
+    public boolean isCompatibleWith(DockerImageName other) {
+        // is this image already the same?
+        final boolean thisRegistrySame = other.registry.equals(this.registry);
+        final boolean thisRepoSame = other.repo.equals(this.repo);
+        final boolean thisVersioningNotSpecifiedOrSame = other.versioning == null ||
+            other.versioning.equals(Versioning.TagVersioning.LATEST) ||
+            other.versioning.equals(this.versioning);
+
+        if (thisRegistrySame && thisRepoSame && thisVersioningNotSpecifiedOrSame) {
+            return true;
         }
 
-        @Override
-        public boolean isValid() {
-            return hash.matches(HASH_REGEX);
+        if (this.compatibleSubstituteFor == null) {
+            return false;
         }
 
-        @Override
-        public String getSeparator() {
-            return "@";
-        }
+        return this.compatibleSubstituteFor.isCompatibleWith(other);
+    }
 
-        @Override
-        public String toString() {
-            return "sha256:" + hash;
+    /**
+     * Behaves as {@link DockerImageName#isCompatibleWith(DockerImageName)} but throws an exception rather than
+     * returning false if a mismatch is detected.
+     *
+     * @param other the other image that we are trying to check compatibility with
+     * @throws IllegalStateException if {@link DockerImageName#isCompatibleWith(DockerImageName)} returns false
+     */
+    public void assertCompatibleWith(DockerImageName other) {
```

Comment:
I think I'd push back against this - it feels a bit strange to have a value be passed through a method that does assertion. I think I like the assertion being a distinct line in each constructor (after the call to `super`), as it feels more visible.

Reply:
The JDK likes to object this argument of feeling strange 🙂 
https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#requireNonNull-T-

But this is not a hill I need to die on. I like it in super constructor though, because this means it gets evaluated before the super constructor is called.